applications/grpc_wallet/src/grpc_server.rs

```diff
@@ -0,0 +1,496 @@
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::grpc_server::wallet_rpc::{
+    server,
+    Contact as ContactRpc,
+    Contacts as ContactsRpc,
+    ReceivedTextMessage as ReceivedTextMessageRpc,
+    RpcResponse,
+    ScreenName as ScreenNameRpc,
+    SentTextMessage as SentTextMessageRpc,
+    TextMessageToSend as TextMessageToSendRpc,
+    TextMessagesResponse as TextMessagesResponseRpc,
+    VoidParams,
+};
+use derive_error::Error;
+use futures::{future, Future, Stream};
+use log::*;
+use std::{convert::TryFrom, net::AddrParseError, sync::Arc};
+use tari_comms::{connection::NetAddress, peer_manager::Peer, types::CommsPublicKey};
+use tari_utilities::message_format::{MessageFormat, MessageFormatError};
+use tari_wallet::{
+    text_message_service::{Contact, TextMessage, TextMessages},
+    Wallet,
+};
+use tokio::net::TcpListener;
+use tower_grpc::{Request, Response};
+use tower_hyper::server::{Http, Server};
+
+pub mod wallet_rpc {
+    include!(concat!(env!("OUT_DIR"), "/wallet_rpc.rs"));
+}
+
+const LOG_TARGET: &'static str = "applications::grpc_wallet";
+
+#[derive(Debug, Error)]
+pub enum WalletServerError {
+    AddrParseError(AddrParseError),
+    IoError(std::io::Error),
+    MessageFormatError(MessageFormatError),
+}
+
+impl TryFrom<TextMessage> for ReceivedTextMessageRpc {
+    type Error = WalletServerError;
+
+    fn try_from(m: TextMessage) -> Result<Self, Self::Error> {
+        Ok(ReceivedTextMessageRpc {
+            id: m.id,
+            source_pub_key: m.source_pub_key.to_base64()?,
+            dest_pub_key: m.dest_pub_key.to_base64()?,
+            message: m.message,
+            timestamp: m.timestamp.to_string(),
+        })
+    }
+}
+
+#[derive(Clone)]
+struct WalletRPC {
+    wallet: Arc<Wallet>,
+}
+
+/// Implementation of the the gRPC service methods.
+impl server::WalletRpc for WalletRPC {
+    type AddContactFuture = future::FutureResult<Response<RpcResponse>, tower_grpc::Status>;
+    type GetContactsFuture = future::FutureResult<Response<ContactsRpc>, tower_grpc::Status>;
```

Comment:
nit. Some of these return structs are called `*****Rpc` and others are `Rpc******`

Reply:
Only RpcResponse is the other way around because it is not the RPC version of a struct but it represents an Rpc Response. The rest are alias' for the Rpc version of a specific struct.